### ISSUE TITLE
Doesnt handle daylight savings

### ISSUE THREAD

**Issue by calumbrodie** [usemarkup/JobQueueBundle#18]
Currently if a recurring job is scheduled to be run at 1:30AM (for example) and daylight savings moves the time back at this time at 2am - it will run twice.

Similarly a job scheduled to run at the same time, wont run when the clocks move forward.

**Comment by gsdevme** [comment]
Are you sure this happens? My thought was crontab handles this edge case

**Comment by calumbrodie** [comment]
but crontab just runs every minute in this bundle. crontab does actually handle this properly but seeing as this bundle replaces crontab it should try and handle this...

**Comment by gsdevme** [comment]
Crontab does not have to be timezone aware, infact the default is for crontab to run in UTC mode just many linux vendors change it (as per my understand). 

Simple solution would be to restore crontab to run in UTC.

**Comment by calumbrodie** [comment]
Not sure you understand the issue. The way this bundle works is by running a job every minute (via crontab).

https://github.com/usemarkup/JobQueueBundle/blob/master/Command/AddRecurringConsoleJobToQueueCommand.php

**Comment by gsdevme** [comment]
Doh "This isn't correct" moment here... I see. Should the JobQueueBundler perhaps do its calculations in UTC then? You specific a timezone say Europe/London to the bundle and it converts the yml into UTC then does a check?

**Comment by calumbrodie** [comment]
@gsdevme the issue is that the 'isDue' will just use the timezone settings of PHP which will change the time when the time changes.

**Comment by gsdevme** [comment]
Yeah `isDue()` needs to calculate in UTC and the yml is in local time and converted to UTC

**Comment by gsdevme** [comment]
I mean like this https://3v4l.org/CWaIZ not sure if Im being clear

**Comment by calumbrodie** [comment]
But it would require not using Cron/CronExpression because it has no support to pass in TimeZone...

**Comment by gsdevme** [comment]
https://github.com/mtdowling/cron-expression/blob/master/src/Cron/CronExpression.php#L276`

```
    public function isDue($currentTime = 'now')
    {
        if ('now' === $currentTime) {
            $currentDate = date('Y-m-d H:i');
            $currentTime = strtotime($currentDate);
        } elseif ($currentTime instanceof DateTime) {
            $currentDate = clone $currentTime;
            // Ensure time in 'current' timezone is used
            $currentDate->setTimezone(new DateTimeZone(date_default_timezone_get()));
            $currentDate = $currentDate->format('Y-m-d H:i');
            $currentTime = strtotime($currentDate);
        } elseif ($currentTime instanceof DateTimeImmutable) {
            $currentDate = DateTime::createFromFormat('U', $currentTime->format('U'));
            $currentDate->setTimezone(new DateTimeZone(date_default_timezone_get()));
            $currentDate = $currentDate->format('Y-m-d H:i');
            $currentTime = strtotime($currentDate);
        } else {
            $currentTime = new DateTime($currentTime);
            $currentTime->setTime($currentTime->format('H'), $currentTime->format('i'), 0);
            $currentDate = $currentTime->format('Y-m-d H:i');
            $currentTime = $currentTime->getTimeStamp();
        }
        try {
            return $this->getNextRunDate($currentDate, 0, true)->getTimestamp() == $currentTime;
        } catch (Exception $e) {
            return false;
        }
    }
```
Problem is this library isn't doing anything wrong really..
```
            // Ensure time in 'current' timezone is used
            $currentDate->setTimezone(new DateTimeZone(date_default_timezone_get()));
```
29th of March 2016 at 1am happens twice... so his library is telling the truth but for our use case we don't want this... I would say this libraries `isDue()` is not fit for our purpose

**Comment by calumbrodie** [comment]
Would rather encourage a PR to the bundle to allow an optionally different behaviour - like being able to set a DateTimeZone against the cron expression which overrides 'date_default_timezone_get()' if that has been set.

Wouldn't it be better for us to assume all jobs are defined in .yml as UTC - and convert them to the current timezone before sending to IsDue?

**Comment by gsdevme** [comment]
That library is doing nothing wrong though, I doubt the PR would get accepted. 

I think our use case for the library is wrong. The 29th of March 1am happens twice in the UK.. That's a fact.
